### Análise Inicial

<p> A análise inicial cosiste em analisar as estruturas da base de dados.

In [81]:
import pandas as pd 

In [82]:
df = pd.read_excel("base_logistica.xlsx", sheet_name="Dados")

In [83]:
df.shape

(320, 15)

In [84]:
df.head()

,ID_Pedido,Data_Pedido,Transportadora,Status_Entrega,Origem,Destino,Categoria_Produto,Modalidade,Peso_KG,Volume_M3,Valor_NF_BRL,Prazo_Contratado_Dias,Dias_Reais_Entrega,Custo_Frete_BRL,Satisfacao_Cliente
0,PED-00001,2024-11-23,Fedex,Atrasado,Rio de Janeiro,Campinas,Móveis,Ferroviário,18.79,1.103,350.06,19,NaN,120.32,5.0
1,PED-00002,2024-02-27,Total Express,Em Trânsito,São Paulo,Goiânia,Eletrônicos,Multimodal,47.54,NaN,4865.08,4,NaN,33.68,5.0
2,PED-00003,2024-01-13,Fedex,Devolvido,Belo Horizonte,Campinas,Livros,Multimodal,36.63,1.324,3211.15,19,20.0,253.26,2.0
3,PED-00004,2025-01-14,TNT,Em Trânsito,Porto Alegre,Brasília,Automotivo,Ferroviário,29.97,0.567,4946.09,11,14.0,566.70,5.0
4,PED-00005,2024-05-20,Total Express,Extraviado,Porto Alegre,Belém,Vestuário,Rodoviário,NaN,1.910,1840.44,8,8.0,169.46,NaN


In [85]:
df.dtypes

ID_Pedido                        object
Data_Pedido              datetime64[ns]
Transportadora                   object
Status_Entrega                   object
Origem                           object
Destino                          object
Categoria_Produto                object
Modalidade                       object
Peso_KG                         float64
Volume_M3                       float64
Valor_NF_BRL                    float64
Prazo_Contratado_Dias             int64
Dias_Reais_Entrega              float64
Custo_Frete_BRL                 float64
Satisfacao_Cliente              float64
dtype: object

In [86]:
df.isnull().sum()#Soma a quantidade de valores nulos na tabela!

ID_Pedido                 0
Data_Pedido               0
Transportadora           25
Status_Entrega            0
Origem                    0
Destino                   0
Categoria_Produto         0
Modalidade               32
Peso_KG                  35
Volume_M3                28
Valor_NF_BRL              0
Prazo_Contratado_Dias     0
Dias_Reais_Entrega       32
Custo_Frete_BRL          41
Satisfacao_Cliente       48
dtype: int64

##### Entendendo os Valores nulos

<p> Valores nulos nem sempre são dados descartáveis como podemos ver, os valores nulos se concentram em colunas chave para a minha análise.<br>
Um desses dados representam a satisfação do cliente que pode nos dizer que:<b>
<ul>
    <li>O cliente ainda não recebeu a mercadoria.</li>
    <li>A mercadoria não foi coletada.</li>
    <li>O cliente nao respondeu a pesquisa de satisfação</li>
</ul>

In [87]:
(df.isnull().sum() / len(df) * 100).round(1) # Calculando a porcentagem de valores nulos em relação aos demais registros na tabela.

ID_Pedido                 0.0
Data_Pedido               0.0
Transportadora            7.8
Status_Entrega            0.0
Origem                    0.0
Destino                   0.0
Categoria_Produto         0.0
Modalidade               10.0
Peso_KG                  10.9
Volume_M3                 8.8
Valor_NF_BRL              0.0
Prazo_Contratado_Dias     0.0
Dias_Reais_Entrega       10.0
Custo_Frete_BRL          12.8
Satisfacao_Cliente       15.0
dtype: float64

#### Conhecendo os valores unicos

<p> A ideia aqui é conhecer os valores na tabela, transportadora, status etc.

In [88]:
df["Transportadora"].value_counts()

Transportadora
TNT              46
Total Express    42
Azul Cargo       38
Fedex            37
Correios         37
Jadlog           33
Sequoia          31
Latam Cargo      31
Name: count, dtype: int64

In [89]:
df["Status_Entrega"].value_counts()

Status_Entrega
Aguardando Coleta    56
Em Trânsito          54
Devolvido            54
Atrasado             53
Extraviado           52
Entregue             51
Name: count, dtype: int64

In [90]:
df["Categoria_Produto"].value_counts()

Categoria_Produto
Automotivo      50
Livros          45
Alimentos       44
Vestuário       44
Móveis          38
Eletrônicos     36
Cosméticos      33
Medicamentos    30
Name: count, dtype: int64

In [91]:
df["Modalidade"].value_counts()

Modalidade
Ferroviário    75
Multimodal     75
Aéreo          73
Rodoviário     65
Name: count, dtype: int64

#### Capturando a descrição de cada Valor 

<p> O comando describe, já calcula os valores de contagem, média mediana, minima,maxima, desvio padrão etc, nos  dando um resumo estatistico dos valores da tabela.

In [92]:
df.describe().round(2)

,Data_Pedido,Peso_KG,Volume_M3,Valor_NF_BRL,Prazo_Contratado_Dias,Dias_Reais_Entrega,Custo_Frete_BRL,Satisfacao_Cliente
count,320,285.00,292.00,320.00,320.00,288.00,279.00,272.00
mean,2024-08-21 00:09:00,25.04,1.00,7208.44,10.18,13.73,410.65,3.04
min,2024-01-01 00:00:00,0.38,0.03,119.25,1.00,0.00,14.55,1.00
25%,2024-04-22 00:00:00,12.17,0.45,3329.15,5.00,9.00,215.36,2.00
50%,2024-08-22 12:00:00,25.86,1.01,6962.02,10.50,14.00,426.36,3.00
75%,2024-12-16 12:00:00,38.10,1.51,10880.04,15.00,18.25,591.46,4.00
max,2025-04-24 00:00:00,49.50,2.00,14953.27,19.00,28.00,797.39,5.00
std,NaN,14.75,0.60,4271.76,5.40,6.70,226.35,1.43


<p> Realizei uma agregação para entender o desvio padrão ddas categorias com maior valor, podendo entender que as categorias, automotivo, cosmeticos, livros e medicamentos, tem um desvio padrão maior que os demais, podendo ter pedidos de muito baixo valor e pedidos de muito alto valor na mesma tabela.

In [93]:
df.groupby("Categoria_Produto")["Valor_NF_BRL"].agg(
   media =  "mean",
   mediana =  "median",
   desvio_padrao = "std"
).round(2)

,media,mediana,desvio_padrao
Categoria_Produto,,,
Alimentos,6995.11,7126.22,4610.88
Automotivo,7824.72,8726.26,4774.00
Cosméticos,7926.98,8494.37,3583.66
Eletrônicos,6221.83,5801.18,3922.79
Livros,7596.31,6570.09,4452.98
Medicamentos,7630.97,7598.30,4214.50
Móveis,6968.30,6728.85,4175.61
Vestuário,6512.42,6119.04,4032.34


##### Fechando o diagnóstico inicial

In [94]:
print("================================= Resumo da Base de Dados")
print(f"Total de Pedidos: {len(df)}")
print(f"Período: {df["Data_Pedido"].min()} até {df["Data_Pedido"].max()}")
print(f"\nTranportadoras: {df["Transportadora"].unique} únicas")
print(f"Categorias: {df["Categoria_Produto"].unique} únicas")
print(f"Status Possíveis: {df["Status_Entrega"].unique} ")
print(f"\nTipos de entragas:: {df["Modalidade"].unique}  ")
print(f"\nColunas com valores nulos")
print(df.isnull().sum()[df.isnull().sum() > 0])


================================= Resumo da Base de Dados
Total de Pedidos: 320
Período: 2024-01-01 00:00:00 até 2025-04-24 00:00:00

Tranportadoras: <bound method Series.unique of 0              Fedex
1      Total Express
2              Fedex
3                TNT
4      Total Express
           ...      
315            Fedex
316         Correios
317              TNT
318          Sequoia
319              TNT
Name: Transportadora, Length: 320, dtype: object> únicas
Categorias: <bound method Series.unique of 0            Móveis
1       Eletrônicos
2            Livros
3        Automotivo
4         Vestuário
           ...     
315    Medicamentos
316    Medicamentos
317      Automotivo
318     Eletrônicos
319     Eletrônicos
Name: Categoria_Produto, Length: 320, dtype: object> únicas
Status Possíveis: <bound method Series.unique of 0               Atrasado
1            Em Trânsito
2              Devolvido
3            Em Trânsito
4             Extraviado
             ...        
315      

In [95]:
def resumo_base(dataframe):
    print("================================= Resumo da Base de Dados =================================================")
    print(f"Total de Pedidos: {len(dataframe)}")
    print(f"Período: {dataframe['Data_Pedido'].min()} até {dataframe['Data_Pedido'].max()}")
    print(f"\nTranportadoras: {dataframe['Transportadora'].unique()} únicas\n")
    print(f"Categorias: {dataframe['Categoria_Produto'].unique()} únicas")
    print(f"Status Possíveis: {dataframe['Status_Entrega'].unique()} ")
    print(f"\nTipos de entregas:: {dataframe['Modalidade'].unique()}  ")
    print(f"\nColunas com valores nulos")
    print(dataframe.isnull().sum()[dataframe.isnull().sum() > 0])
    print("\n================================= FIM DO RESUMO =================================================")


In [96]:
resumo_base(df)

================================= Resumo da Base de Dados =================================================
Total de Pedidos: 320
Período: 2024-01-01 00:00:00 até 2025-04-24 00:00:00

Tranportadoras: ['Fedex' 'Total Express' 'TNT' 'Jadlog' 'Sequoia' nan 'Correios'
 'Latam Cargo' 'Azul Cargo'] únicas

Categorias: ['Móveis' 'Eletrônicos' 'Livros' 'Automotivo' 'Vestuário' 'Alimentos'
 'Cosméticos' 'Medicamentos'] únicas
Status Possíveis: ['Atrasado' 'Em Trânsito' 'Devolvido' 'Extraviado' 'Entregue'
 'Aguardando Coleta'] 

Tipos de entregas:: ['Ferroviário' 'Multimodal' 'Rodoviário' 'Aéreo' nan]  

Colunas com valores nulos
Transportadora        25
Modalidade            32
Peso_KG               35
Volume_M3             28
Dias_Reais_Entrega    32
Custo_Frete_BRL       41
Satisfacao_Cliente    48
dtype: int64

================================= FIM DO RESUMO =================================================


### Tratando Valores Nulos 

Nesse espaço realizarei as tratativas referentes a valores nulos na tabela entendendo o que cada um deles representa.

In [97]:
df.columns

Index(['ID_Pedido', 'Data_Pedido', 'Transportadora', 'Status_Entrega',
       'Origem', 'Destino', 'Categoria_Produto', 'Modalidade', 'Peso_KG',
       'Volume_M3', 'Valor_NF_BRL', 'Prazo_Contratado_Dias',
       'Dias_Reais_Entrega', 'Custo_Frete_BRL', 'Satisfacao_Cliente'],
      dtype='object')

##### COLUNA TRANSPORTADORA 

In [98]:
df["Transportadora"] = df["Transportadora"].fillna("Não Informado")


In [99]:
df["Transportadora"].isnull().sum()

np.int64(0)

##### COLUNA MODALIDADE

In [100]:
df["Modalidade"] = df.groupby("Categoria_Produto")["Modalidade"].transform(
    lambda x: x.fillna(x.mode()[0] if not x.mode().empty else "Rodoviario")
)

In [101]:
df["Modalidade"].isnull().sum()

np.int64(0)

#### COLUNAS NÚMERICAS 

<p>Aqui a tratativa não sera um valor fixo, preencherei com o calculo da mediana justamente por conta do desvio padrão alto


#### Peso e Volume

In [102]:
df["Peso_KG"] = df["Peso_KG"].fillna(df["Peso_KG"].median())
df["Volume_M3"] = df["Volume_M3"].fillna(df["Volume_M3"].median())

In [103]:
df["Peso_KG"].isnull().sum()
df["Volume_M3"].isnull().sum()

np.int64(0)

#### Dias Reais de Entrega

In [104]:
df["Dias_Reais_Entrega"] = df.groupby("Status_Entrega")["Dias_Reais_Entrega"].transform(
    lambda x:x.fillna(x.median())
)

In [105]:
df["Dias_Reais_Entrega"].isnull().sum()

np.int64(0)

#### Custo Frete BRL

In [106]:
df["Custo_Frete_BRL"] = df["Custo_Frete_BRL"].fillna(df["Custo_Frete_BRL"].median())

#### Satisfacao do CLiente

In [107]:
df["Satisfacao_Cliente"] = df["Satisfacao_Cliente"].fillna(df["Satisfacao_Cliente"].median())

In [108]:
df["Satisfacao_Cliente"].isnull().sum()

np.int64(0)

#### Confirmando se todas as colunas estão com os dados preenchidos

In [109]:
df.isnull().sum()

ID_Pedido                0
Data_Pedido              0
Transportadora           0
Status_Entrega           0
Origem                   0
Destino                  0
Categoria_Produto        0
Modalidade               0
Peso_KG                  0
Volume_M3                0
Valor_NF_BRL             0
Prazo_Contratado_Dias    0
Dias_Reais_Entrega       0
Custo_Frete_BRL          0
Satisfacao_Cliente       0
dtype: int64

#### Criando Coluna para análise de performance com base na entrega

In [110]:
df["Desvio_Dias"] = df["Dias_Reais_Entrega"] - df["Prazo_Contratado_Dias"]

In [111]:
df["Desvio_Dias"].describe().round(2)

count    320.00
mean       3.57
std        3.82
min       -7.00
25%        0.00
50%        3.00
75%        7.00
max       14.00
Name: Desvio_Dias, dtype: float64

In [112]:
df["Pedido_Atrasado"] = (df["Desvio_Dias"] > 0).astype(int)
df["Pedido_Atrasado"].value_counts()

Pedido_Atrasado
1    239
0     81
Name: count, dtype: int64

In [113]:

(df["Pedido_Atrasado"].mean().round(2))

np.float64(0.75)

In [114]:
df["Custo_Por_KG"] = (df["Custo_Frete_BRL"] / df["Peso_KG"]).round(2)


In [115]:
df["Mês"] = df["Data_Pedido"].dt.month
df["Trimestre"] = df["Data_Pedido"].dt.quarter
df["Ano"] = df["Data_Pedido"].dt.year
df["Mês_Ano"] = df["Data_Pedido"].dt.to_period("M").astype(str)

In [116]:
df[["Data_Pedido", "Mês", "Trimestre", "Ano", "Mês_Ano"]].head()


,Data_Pedido,Mês,Trimestre,Ano,Mês_Ano
0,2024-11-23,11,4,2024,2024-11
1,2024-02-27,2,1,2024,2024-02
2,2024-01-13,1,1,2024,2024-01
3,2025-01-14,1,1,2025,2025-01
4,2024-05-20,5,2,2024,2024-05


#### Salvando a base em diferentes formatos

In [117]:
df.to_parquet("base_limpa.parquet", index=False)

In [118]:
df.to_csv("base_limpa.csv", index=False, encoding="utf-8-sig",decimal=",",sep=";")

In [119]:
print(f"Shape Final: {df.shape}")

Shape Final: (320, 22)


In [120]:
resumo_base(df)

================================= Resumo da Base de Dados =================================================
Total de Pedidos: 320
Período: 2024-01-01 00:00:00 até 2025-04-24 00:00:00

Tranportadoras: ['Fedex' 'Total Express' 'TNT' 'Jadlog' 'Sequoia' 'Não Informado'
 'Correios' 'Latam Cargo' 'Azul Cargo'] únicas

Categorias: ['Móveis' 'Eletrônicos' 'Livros' 'Automotivo' 'Vestuário' 'Alimentos'
 'Cosméticos' 'Medicamentos'] únicas
Status Possíveis: ['Atrasado' 'Em Trânsito' 'Devolvido' 'Extraviado' 'Entregue'
 'Aguardando Coleta'] 

Tipos de entregas:: ['Ferroviário' 'Multimodal' 'Rodoviário' 'Aéreo']  

Colunas com valores nulos
Series([], dtype: int64)

================================= FIM DO RESUMO =================================================


#### RESUMO DA LIMPEZA DOS DADOS

<p> Podemos observar que a função que antes retornava o resumo com algumas informações de valores unicos e valores nulos retorna porém com a tratativa feita referente aos valores nulos.<br>
Criei algumas colunas para explorar e complementar a análise por exemplo a coluna de data fracionada por mes, ano, trimestre e mes_ano.<br>
A coluna Desvio_Dias mostra exatamente se o desvio de entrega está positivo opu negativo, positivo sendo ruim (Entrega atrasada), negativo sendo entrega antecipada.</p>
<p> Agora a ideia é partir para a análise exploratória e responder as perguntas de negócios. 


### 📋 Perguntas de Negócio — Base Logística


#### Performance Operacional

<ul>
<li>Qual é a taxa geral de atraso da operação?</li>
<li>Quantos dias em média os pedidos atrasam?</li>
<li>Qual o pior e melhor mês em termos de atraso?</li>
</ul>


#### Transportadoras
<uL>
<li>Qual transportadora tem a maior taxa de atraso?</li>
<li>Qual transportadora tem o melhor custo-benefício (custo × atraso × satisfação)?</li>
<li>Qual transportadora é mais cara por KG transportado?</li>
</ul>


#### Rotas

<ul>
<li>Quais rotas concentram mais atrasos?</li>
<li>Quais origens geram mais pedidos?</li>
<li>Existe alguma rota com custo desproporcional?</li>
</ul>

#### Produtos

<ul>
<li>Quais categorias de produto têm maior valor de NF em risco?</li>
<li>Quais categorias mais sofrem com atrasos?</li>
</ul>

#### Cliente

<ul>
<li>A satisfação do cliente cai quando há atraso?</li>
<li>Qual transportadora tem a pior avaliação dos clientes?</li>
</ul>

#### Financeiro

<ul>
<li>Qual o custo total de frete por transportadora?</li>
<li>O custo do frete é proporcional ao valor da NF?</li>
</uls>

## Análise Exploratoria

In [121]:
status  = df["Status_Entrega"].value_counts()
porcentagem = (status / len(df) * 100).round(2)

In [122]:
pd.DataFrame({"Quantidade": status, "Percentual %": porcentagem} )

,Quantidade,Percentual %
Status_Entrega,,
Aguardando Coleta,56,17.50
Em Trânsito,54,16.88
Devolvido,54,16.88
Atrasado,53,16.56
Extraviado,52,16.25
Entregue,51,15.94


In [123]:
df.columns

Index(['ID_Pedido', 'Data_Pedido', 'Transportadora', 'Status_Entrega',
       'Origem', 'Destino', 'Categoria_Produto', 'Modalidade', 'Peso_KG',
       'Volume_M3', 'Valor_NF_BRL', 'Prazo_Contratado_Dias',
       'Dias_Reais_Entrega', 'Custo_Frete_BRL', 'Satisfacao_Cliente',
       'Desvio_Dias', 'Pedido_Atrasado', 'Custo_Por_KG', 'Mês', 'Trimestre',
       'Ano', 'Mês_Ano'],
      dtype='object')

##### Qual é a taxa geral de atraso da operação?
#### Quantos dias em média os pedidos atrasam?

In [148]:
mensal= df.groupby("Mês_Ano").agg(
    Pedidos = ("ID_Pedido","count"),
    Atrasos = ("Pedido_Atrasado", "sum")
).reset_index()


mensal["Atrasos"] = mensal["Atrasos"].astype(float)
mensal["Pedidos"] = mensal["Pedidos"].astype(float)


In [149]:
mensal["Taxa_Atraso_%"] = (mensal["Atrasos"] / mensal["Pedidos"] * 100).round(1)


In [150]:
mensal = mensal.sort_values("Atrasos",ascending=False)

In [151]:
mensal

,Mês_Ano,Pedidos,Atrasos,Taxa_Atraso_%
3,2024-04,24.0,20.0,83.3
2,2024-03,20.0,18.0,90.0
4,2024-05,25.0,18.0,72.0
10,2024-11,23.0,17.0,73.9
12,2025-01,26.0,17.0,65.4
1,2024-02,27.0,16.0,59.3
0,2024-01,19.0,16.0,84.2
7,2024-08,17.0,15.0,88.2
14,2025-03,21.0,14.0,66.7
6,2024-07,18.0,14.0,77.8


#### Análise por transportadora

In [128]:
transportadoras = df.groupby("Transportadora").agg(
    Pedidos = ("ID_Pedido", "count"),
    Desvio_Medio= ("Desvio_Dias", "mean"),
    Satisfacao = ("Satisfacao_Cliente", "mean"),
    Custo_Medio = ("Custo_Frete_BRL", "mean"),
    Custo_Por_KG = ("Custo_Por_KG", "mean") 

).round(2)

In [129]:
transportadoras["Taxa_Atraso_%"] = (df.groupby("Transportadora")["Pedido_Atrasado"].mean() * 100).round(1)
transportadoras = transportadoras.sort_values("Taxa_Atraso_%", ascending=False)

In [130]:
transportadoras

,Pedidos,Desvio_Medio,Satisfacao,Custo_Medio,Custo_Por_KG,Taxa_Atraso_%
Transportadora,,,,,,
Azul Cargo,38,4.63,2.95,433.08,29.37,86.8
Não Informado,25,4.16,2.84,400.42,18.88,80.0
Correios,37,2.92,2.95,393.20,34.30,75.7
Fedex,37,3.73,2.84,394.19,31.22,75.7
Sequoia,31,4.10,2.87,414.11,29.40,74.2
TNT,46,3.22,3.02,407.11,65.74,73.9
Total Express,42,3.48,3.02,436.51,33.78,73.8
Latam Cargo,31,3.26,3.48,394.07,47.17,67.7
Jadlog,33,2.85,3.39,434.51,115.50,63.6


In [131]:
transportadoras[["Pedidos", "Taxa_Atraso_%", "Custo_Por_KG", "Satisfacao"]].sort_values("Satisfacao", ascending=False)

,Pedidos,Taxa_Atraso_%,Custo_Por_KG,Satisfacao
Transportadora,,,,
Latam Cargo,31,67.7,47.17,3.48
Jadlog,33,63.6,115.50,3.39
TNT,46,73.9,65.74,3.02
Total Express,42,73.8,33.78,3.02
Azul Cargo,38,86.8,29.37,2.95
Correios,37,75.7,34.30,2.95
Sequoia,31,74.2,29.40,2.87
Fedex,37,75.7,31.22,2.84
Não Informado,25,80.0,18.88,2.84


In [132]:
df[df["Transportadora"] == "Latam Cargo"][["Prazo_Contratado_Dias", "Desvio_Dias", "Satisfacao_Cliente"]].describe().round(2)

,Prazo_Contratado_Dias,Desvio_Dias,Satisfacao_Cliente
count,31.00,31.00,31.00
mean,10.03,3.26,3.48
std,5.01,3.80,1.29
min,1.00,-2.00,1.00
25%,6.00,0.00,3.00
50%,10.00,3.00,3.00
75%,14.00,6.50,5.00
max,18.00,9.00,5.00


In [133]:
df[df["Transportadora"] == "Latam Cargo"]["Prazo_Contratado_Dias"].describe().round(2)

count    31.00
mean     10.03
std       5.01
min       1.00
25%       6.00
50%      10.00
75%      14.00
max      18.00
Name: Prazo_Contratado_Dias, dtype: float64

In [134]:
df[df["Transportadora"] == "Azul Cargo"]["Prazo_Contratado_Dias"].describe().round(2)

count    38.00
mean      9.84
std       5.12
min       2.00
25%       5.00
50%      11.00
75%      13.75
max      18.00
Name: Prazo_Contratado_Dias, dtype: float64

In [135]:
df[df["Transportadora"] == "Latam Cargo"]["Status_Entrega"].value_counts()

Status_Entrega
Em Trânsito          8
Aguardando Coleta    7
Entregue             7
Extraviado           5
Atrasado             4
Name: count, dtype: int64

In [136]:
df[df["Transportadora"] == "Azul Cargo"]["Status_Entrega"].value_counts()

Status_Entrega
Extraviado           7
Em Trânsito          7
Entregue             7
Atrasado             7
Aguardando Coleta    6
Devolvido            4
Name: count, dtype: int64

### 💡 Insight 1 — Transportadoras
A Latam Cargo tem alta satisfação apesar dos atrasos pois preserva 
a mercadoria — poucos extravios e devoluções. Já a Azul Cargo combina 
atraso + extravio, o que explica sua menor satisfação.

**Recomendação:** revisar contrato com Azul Cargo e investigar 
processo de rastreamento de cargas.

#### Transportadoras

In [137]:
rotas = df.groupby(["Origem", "Destino"]).agg(
    Pedidos      = ("ID_Pedido", "count"),
    Atraso_Pct   = ("Pedido_Atrasado", "mean"),
    Desvio_Medio = ("Desvio_Dias", "mean"),
    Custo_Medio  = ("Custo_Frete_BRL", "mean"),
).round(2)

rotas["Atraso_Pct"] = (rotas["Atraso_Pct"] * 100).round(1)

rotas.sort_values("Pedidos", ascending=False).head(10)

Pedidos  Atraso_Pct  Desvio_Medio  Custo_Medio
Origem         Destino                                                       
Belo Horizonte Ribeirão Preto        9        67.0          3.78       349.24
Porto Alegre   Manaus                9        67.0          3.22       447.76
Rio de Janeiro Belém                 9        89.0          4.44       394.53
Recife         Campinas              9        56.0          1.89       465.97
Fortaleza      Campinas              8        88.0          3.50       282.02
               Natal                 8        88.0          3.88       381.98
               Belém                 8        62.0          2.88       415.84
Curitiba       Campinas              8        75.0          5.50       394.29
Salvador       Campinas              8        88.0          4.25       443.62
Rio de Janeiro Ribeirão Preto        8        88.0          4.88       531.90

Qual é a rota com mais pedidos?

<p> A rota com mais pedidos estão circulando entre as 4:<br> 
BH - RP<br>
Porto Alegr - Manaus<br>
RJ - Belém <br>
Recife - Campinas<br>



<p>Qual rota tem a maior taxa de atraso?


<p>A rota com a maior taxa de é RJ - Belém</p>
<p>Porém a rota mais cara e com mais atraso é RJ - Ribeirão Preto.<br>
Pois tem uma Taxa de atraso altissima  um desvio de quase 5 diase um custo médio de R$ 531,90

In [138]:
df[(df["Origem"] == "Rio de Janeiro") & 
   (df["Destino"] == "Ribeirão Preto")][["Transportadora", "Modalidade", "Pedido_Atrasado", "Desvio_Dias", "Custo_Frete_BRL"]].sort_values("Desvio_Dias", ascending=False)

,Transportadora,Modalidade,Pedido_Atrasado,Desvio_Dias,Custo_Frete_BRL
46,Azul Cargo,Ferroviário,1,10.0,656.09
137,Latam Cargo,Aéreo,1,9.0,426.36
182,Não Informado,Ferroviário,1,9.0,795.37
250,TNT,Ferroviário,1,7.0,114.69
50,Fedex,Rodoviário,1,3.0,266.42
150,Latam Cargo,Multimodal,1,1.0,790.35
266,Fedex,Multimodal,1,1.0,773.00
316,Correios,Aéreo,0,-1.0,432.90


In [139]:
df[(df["Origem"] == "Rio de Janeiro") & 
   (df["Destino"] == "Ribeirão Preto")]["Transportadora"].value_counts()

Transportadora
Fedex            2
Latam Cargo      2
Azul Cargo       1
Não Informado    1
TNT              1
Correios         1
Name: count, dtype: int64

### 💡 Insight 2 — Rotas
A rota RJ → Ribeirão Preto é a mais cara e com alto atraso.
Todas as transportadoras que operam nela atrasam — indicando 
gargalo estrutural da rota, não operacional.

**Recomendação:** avaliar rota alternativa via modal diferente
ou hub intermediário para essa origem-destino.

In [140]:
categorias = df.groupby("Categoria_Produto").agg(
    Pedidos     = ("ID_Pedido", "count"),
    Valor_Total = ("Valor_NF_BRL", "sum"),
    Atraso_Pct  = ("Pedido_Atrasado", "mean"),
    Satisfacao  = ("Satisfacao_Cliente", "mean"),
).round(2)

categorias["Atraso_Pct"] = (categorias["Atraso_Pct"] * 100).round(1)
categorias["Valor_Total"] = categorias["Valor_Total"].round(0)

categorias.sort_values("Valor_Total", ascending=False)

,Pedidos,Valor_Total,Atraso_Pct,Satisfacao
Categoria_Produto,,,,
Automotivo,50,391236.0,74.0,3.16
Livros,45,341834.0,73.0,3.16
Alimentos,44,307785.0,77.0,2.73
Vestuário,44,286547.0,75.0,2.91
Móveis,38,264795.0,79.0,3.11
Cosméticos,33,261590.0,76.0,3.27
Medicamentos,30,228929.0,63.0,3.10
Eletrônicos,36,223986.0,78.0,2.92


In [141]:
categorias.sort_values("Atraso_Pct", ascending=False)

,Pedidos,Valor_Total,Atraso_Pct,Satisfacao
Categoria_Produto,,,,
Móveis,38,264795.0,79.0,3.11
Eletrônicos,36,223986.0,78.0,2.92
Alimentos,44,307785.0,77.0,2.73
Cosméticos,33,261590.0,76.0,3.27
Vestuário,44,286547.0,75.0,2.91
Automotivo,50,391236.0,74.0,3.16
Livros,45,341834.0,73.0,3.16
Medicamentos,30,228929.0,63.0,3.10


In [142]:
df[(df["Categoria_Produto"] == "Alimentos") & 
   (df["Status_Entrega"] == "Extraviado")].agg(
    Pedidos_Extraviados = ("ID_Pedido", "count"),
    Valor_Extraviado    = ("Valor_NF_BRL", "sum"),
    Custo_Frete_Perdido = ("Custo_Frete_BRL", "sum"),
).round(2)

,ID_Pedido,Valor_NF_BRL,Custo_Frete_BRL
Pedidos_Extraviados,10.0,NaN,NaN
Valor_Extraviado,NaN,66203.67,NaN
Custo_Frete_Perdido,NaN,NaN,5963.67


In [143]:
df[df["Status_Entrega"] == "Extraviado"].agg(
    Pedidos_Extraviados = ("ID_Pedido", "count"),
    Valor_Extraviado    = ("Valor_NF_BRL", "sum"),
    Custo_Frete_Perdido = ("Custo_Frete_BRL", "sum"),
).round(2)

,ID_Pedido,Valor_NF_BRL,Custo_Frete_BRL
Pedidos_Extraviados,52.0,NaN,NaN
Valor_Extraviado,NaN,328633.36,NaN
Custo_Frete_Perdido,NaN,NaN,22924.48


### 💡 Insight 3 — Prejuízo com Extravios
R$ 328.633,36 em mercadorias extraviadas + R$ 22.924,48 em 
frete pago e não entregue = R$ 351.557,84 de prejuízo direto.

Categoria Alimentos concentra R$ 72.167,34 desse total —
produto perecível com risco sanitário adicional.

**Recomendação:** implementar seguro de carga obrigatório e 
rastreamento em tempo real para todas as entregas.

In [144]:
df.groupby("Status_Entrega")["Satisfacao_Cliente"].agg(
    Media  = "mean",
    Minimo = "min",
    Maximo = "max"
).round(2).sort_values("Media", ascending=False)

,Media,Minimo,Maximo
Status_Entrega,,,
Em Trânsito,3.20,1.0,5.0
Devolvido,3.09,1.0,5.0
Atrasado,3.04,1.0,5.0
Aguardando Coleta,2.98,1.0,5.0
Extraviado,2.96,1.0,5.0
Entregue,2.94,1.0,5.0


In [145]:
entregues = df[df["Status_Entrega"] == "Entregue"].copy()

entregues.groupby("Pedido_Atrasado")["Satisfacao_Cliente"].agg(
    Media   = "mean",
    Mediana = "median",
    Count   = "count"
).round(2)

,Media,Mediana,Count
Pedido_Atrasado,,,
0,3.09,3.0,11
1,2.90,3.0,40


### 💡 Insight 4 — Satisfação x Atraso
Pedidos entregues no prazo têm satisfação média de 3.09
contra 2.90 dos entregues com atraso.

Clientes em status intermediários (Aguardando Coleta, Em Trânsito)
avaliam melhor por ainda terem expectativa positiva.

**Recomendação:** comunicar proativamente atrasos ao cliente —
a surpresa negativa impacta mais do que o atraso em si.

In [146]:
financeiro = df.groupby("Transportadora").agg(
    Custo_Total  = ("Custo_Frete_BRL", "sum"),
    Valor_NF     = ("Valor_NF_BRL", "sum"),
).round(2)

financeiro["Pct_Frete_NF"] = (financeiro["Custo_Total"] / financeiro["Valor_NF"] * 100).round(2)

financeiro.sort_values("Custo_Total", ascending=False)

,Custo_Total,Valor_NF,Pct_Frete_NF
Transportadora,,,
TNT,18726.98,329855.30,5.68
Total Express,18333.33,308852.40,5.94
Azul Cargo,16456.89,348571.18,4.72
Fedex,14585.00,251772.65,5.79
Correios,14548.42,245376.65,5.93
Jadlog,14338.78,231062.73,6.21
Sequoia,12837.39,250534.02,5.12
Latam Cargo,12216.20,188588.35,6.48
Não Informado,10010.47,152088.84,6.58


### 💡 Insight 5 — Financeiro
Azul Cargo transporta o maior valor em mercadorias (R$ 348k)
com o menor custo relativo de frete (4.72%) — porém é a 
transportadora com maior taxa de atraso e extravios.

Latam Cargo tem o maior custo relativo (6.48%) mas melhor
satisfação — o custo extra pode ser justificado pela qualidade.

**Recomendação:** redistribuir cargas de alto valor da Azul Cargo
para transportadoras com menor índice de extravio.

In [ ]:
df.columns